certora/specs/AccessManager.spec

```diff
@@ -0,0 +1,588 @@
+*/
+invariant publicRole(env e, address account)
+    hasRole_1(e, PUBLIC_ROLE(), account)        &&
+    hasRole_2(e, PUBLIC_ROLE(), account)   == 0 &&
+    getAccess_1(e, PUBLIC_ROLE(), account) == 0 &&
+    getAccess_2(e, PUBLIC_ROLE(), account) == 0 &&
+    getAccess_3(e, PUBLIC_ROLE(), account) == 0 &&
+    getAccess_4(e, PUBLIC_ROLE(), account) == 0;
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: hasRole is consistent with getAccess                                                                     │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant hasRoleGetAccessConsistency(env e, uint64 roleId, address account)
+    hasRole_1(e, roleId, account) == (roleId == PUBLIC_ROLE() || isSetAndPast(e, getAccess_1(e, roleId, account))) &&
+    hasRole_2(e, roleId, account) == getAccess_2(e, roleId, account);
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Functions: canCall, getAccess and hasRole do NOT revert                                                             │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule noRevert(env e) {
+    require nonpayable(e);
+    require envsanity(e);
+
+    address caller;
+    address target;
+    bytes4 selector;
+    uint64 roleId;
+
+    canCall@withrevert(e, caller, target, selector);
+    assert !lastReverted;
+
+    getAccess@withrevert(e, roleId, caller);
+    assert !lastReverted;
+
+    hasRole@withrevert(e, roleId, caller);
+    assert !lastReverted;
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Functions: canCall                                                                                                  │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule canCall(env e) {
+    address caller;
+    address target;
+    bytes4 selector;
+
+    // Get relevant values
+    bool   immediate    = canCall_1(e, caller, target, selector);
+    uint32 delay        = canCall_2(e, caller, target, selector);
+    bool   closed       = isTargetClosed(target);
+    uint64 roleId       = getTargetFunctionRole(target, selector);
+    bool   isMember     = hasRole_1(e, roleId, caller);
+    uint32 currentDelay = hasRole_2(e, roleId, caller);
+
+    // Can only execute without delay in specific cases:
+    // - target not closed
+    // - if self-execution: `executionId` must match
+    // - if third party execution: must be member with no delay
+    assert immediate <=> (
+        !closed &&
+        (
+            (caller == currentContract && executionId() == hashExecutionId(target, selector))
```

Comment:
What happens if we remove this line?

I initially thought it's ok because we've seen CVL starting to evaluate a rule with data already in storage, but I wonder why the `cleanExecutionId()` rule is working then, am I missing something?

Reply:
`cleanExecutionId()` is an invariant that verifies that in between calls the executionId is always 0. We should interpret this has "value is cleaned at the end of tx"

This rule doesn't make such assumption. We could require the invariant, but that would be a mistake on our part because it would not consider cases where `canCall` is called in the middle of a transaction happening (inside an execute).

Then if we remove this line, we would get an error saying: 
- if the caller is the current contract
- if canCall is called in a context where executionId is set
- if the executionId matched the `hashExecutionId` for that call
then you'd be able to call immediatly.

→ that is something we want.